### PR TITLE
Create converter for ArcGis FeatureServer type classBreaks

### PR DESCRIPTION
**Pull request by merydian**
## Description

The file qgsarcgisrestutils.cpp is currently not supporting "classBreaks" type ArcGis-FeatureServers. This pull request adds functionality for QGIS to automatically convert classBreaks FeatureServer into a QgsGraduatedSymbolRenderer. Symbology will be loaded automatically now. 

image

### CODE REVIEW

File: tests/src/python/test_provider_afs.py [hunk 1]
```diff
@@ -1223,6 +1225,178 @@ def testRenderer(self):
...
+         ]
+        }
+        """)
+
+        # Create test layer
+        vl = QgsVectorLayer("url='http://" + endpoint + "' crs='epsg:3857'", 'test', 'arcgisfeatureserver')
+        self.assertTrue(vl.isValid())
+        self.assertIsNotNone(vl.dataProvider().createRenderer())
+        self.assertIsInstance(vl.renderer(), QgsGraduatedSymbolRenderer)
+        self.assertIsInstance(vl.renderer().sourceSymbol(), QgsSymbol)
```
Comment (nyalldawson) [inline tests/src/python/test_provider_afs.py:1397]: Can you also test for the ranges created here? Eg check the count of ranges for the renderer, and their min/max values?

**Comment by nyalldawson** [comment]
Nice work @merydian ! 

In order to fix the failing indentation tests you'll need to run the "scripts/astyle_all.sh" script over your branch. That'll auto-apply the required indentation and flake formatting.

(It's easiest to set this up as a pre-commit hook using the scripts/prepare_commit.sh script, so that all the formatting and such is automatically applied for you!)

File: src/core/providers/arcgis/qgsarcgisrestutils.cpp [hunk 2]
```diff
@@ -945,8 +952,122 @@ QgsFeatureRenderer *QgsArcGisRestUtils::convertRenderer( const QVariantMap &rend
...
+                  graduatedRenderer->addClass( range );
+              }
+          }
+      }
+      lastValue = rendererData.value( QStringLiteral( "minValue" ) ).toDouble();
+      for ( const QVariant& classBreakInfo : classBreakInfos )
+      {
+          const QVariantMap symbolData = classBreakInfo.toMap().value( QStringLiteral( "symbol" ) ).toMap();
+          std::unique_ptr< QgsSymbol > symbol( QgsArcGisRestUtils::convertSymbol( symbolData ) );
+          double classMaxValue = classBreakInfo.toMap().value( QStringLiteral( "classMaxValue" ) ).toFloat();
```
Comment (nyalldawson) [inline src/core/providers/arcgis/qgsarcgisrestutils.cpp:1055]: Missed one here:

```suggestion
          double classMaxValue = classBreakInfo.toMap().value( QStringLiteral( "classMaxValue" ) ).toDouble();
```

File: tests/src/python/test_provider_afs.py (hunk 1 above)
Comment (merydian) [reply tests/src/python/test_provider_afs.py:1399]: Done

File: src/providers/arcgisrest/qgsafsprovider.cpp [hunk 3]
```diff
@@ -313,6 +313,8 @@ QgsAfsProvider::QgsAfsProvider( const QString &uri, const ProviderOptions &optio
   // renderer
   mRendererDataMap = layerData.value( QStringLiteral( "drawingInfo" ) ).toMap().value( QStringLiteral( "renderer" ) ).toMap();
   mLabelingDataList = layerData.value( QStringLiteral( "drawingInfo" ) ).toMap().value( QStringLiteral( "labelingInfo" ) ).toList();
+  const double transparency = layerData.value(QStringLiteral("drawingInfo")).toMap().value(QStringLiteral("transparency")).toDouble();
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsafsprovider.cpp:316]:
```suggestion
  const double transparency = layerData.value( QStringLiteral( "drawingInfo" ) ).toMap().value( QStringLiteral( "transparency" ) ).toDouble();
```

File: src/providers/arcgisrest/qgsafsprovider.cpp [hunk 4]
```diff
@@ -313,6 +313,8 @@ QgsAfsProvider::QgsAfsProvider( const QString &uri, const ProviderOptions &optio
...
+  mRendererDataMap.insert(QStringLiteral("transparency"), QVariant(transparency));
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsafsprovider.cpp:317]:
```suggestion
  mRendererDataMap.insert( QStringLiteral( "transparency" ), QVariant( transparency ) );
```

File: src/core/providers/arcgis/qgsarcgisrestutils.cpp [hunk 5]
```diff
@@ -945,8 +952,131 @@ QgsFeatureRenderer *QgsArcGisRestUtils::convertRenderer( const QVariantMap &rend
...
-    // currently unsupported
-    return nullptr;
+    const QString attrName = rendererData.value( QStringLiteral( "field" ) ).toString();
+    QgsGraduatedSymbolRenderer *graduatedRenderer = new QgsGraduatedSymbolRenderer( attrName );
+
+    const QVariantList classBreakInfos = rendererData.value( QStringLiteral( "classBreakInfos" ) ).toList();
+    const QVariantMap authoringInfo = rendererData.value( QStringLiteral( "authoringInfo" ) ).toMap();
+    QVariantMap symbolData;
+
+    QString esriMode = rendererData.value( QStringLiteral( "authoringInfo" ) ).toMap().value( QStringLiteral( "classificationMethod" ) ).toString();
```
Comment (nyalldawson) [inline src/core/providers/arcgis/qgsarcgisrestutils.cpp:962]:
```suggestion
    QString esriMode = authoringInfo.value( QStringLiteral( "classificationMethod" ) ).toString();
```

File: src/core/providers/arcgis/qgsarcgisrestutils.cpp [hunk 6]
```diff
@@ -945,8 +952,131 @@ QgsFeatureRenderer *QgsArcGisRestUtils::convertRenderer( const QVariantMap &rend
...
+      graduatedRenderer->setSourceSymbol( symbol.release() );
+    }
+
+    const QVariantList visualVariablesData = rendererData.value( QStringLiteral( "visualVariables" ) ).toList();
+    double lastValue = rendererData.value( QStringLiteral( "minValue" ) ).toDouble();
+    for ( const QVariant &visualVariable : visualVariablesData )
+    {
+      const QVariantList stops = visualVariable.toMap().value( QStringLiteral( "stops" ) ).toList();
+      QString lastLabel;
+
```
Comment (nyalldawson) [inline src/core/providers/arcgis/qgsarcgisrestutils.cpp:1033]:
```suggestion
```

File: tests/src/python/test_provider_afs.py [hunk 7]
```diff
@@ -1223,6 +1226,182 @@ def testRenderer(self):
         self.assertEqual(vl.renderer().categories()[0].value(), 'US')
         self.assertEqual(vl.renderer().categories()[1].value(), 'Canada')
 
+    def testGraduatedRenderer(self):
+        """ Test that the graduated renderer is correctly acquired from provider """
+
+        endpoint = self.basetestpath + '/renderer_fake_qgis_http_endpoint'
```
Comment (nyalldawson) [inline tests/src/python/test_provider_afs.py:1232]:
```suggestion
        endpoint = self.basetestpath + '/class_breaks_renderer_fake_qgis_http_endpoint'
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Nice work @merydian !

**Pull request merged**